### ISSUE TITLE
Digital Ocean doesn't work with dopy 0.3.7

### ISSUE THREAD

**Issue by Paxa** [ansible/ansible-modules-core#2509]
##### ISSUE TYPE

Bug Report
##### COMPONENT NAME

digital_ocean module
##### ANSIBLE VERSION

N/A
##### SUMMARY

When I do `sudo pip instal dopy` it will instal dopy 0.3.7

And then when I try to use DigitalOcean module it will show (running with `-vvvv`)

```
localhost EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/env python /Users/pavel/.ansible/tmp/ansible-tmp-1448140256.97-175424737400186/digital_ocean; rm -rf "/Users/pavel/.ansible/tmp/ansible-tmp-1448140256.97-175424737400186/" > /dev/null 2>&1
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/Users/pavel/.ansible/tmp/ansible-tmp-1448140256.97-175424737400186/digital_ocean", line 182, in <module>
    class TimeoutError(DoError):
NameError: name 'DoError' is not defined

```

But when I install dopy 0.3.5 then it works well (`sudo pip install 'dopy>=0.3.5,<=0.3.5'`)

**Comment by zubrabubra** [comment]
@Paxa hello!
You might need to install six(`sudo pip install six`) package, then upgrade dopy to the latest version and double check.

**Comment by CameronCarranza** [comment]
I've run into this issue recently as well runnning Ansible 2.0 off of the Stable branch as well as devel. Switching back to 0.3.5 as @Paxa said allows me to run the playbook just fine, but 0.3.7 results in the same exact error.

I have six 1.10.0 installed as well. Issue is affecting me on OS X El Capitan running Python 2.7 via HomeBrew.

**Comment by CraigMason** [comment]
Confirmed I'm seeing the same issue on OSX 10.11 (`ansible` 1.9.4, installed in virtualenv, with `dopy` 0.3.7). I'm using a copied version of the DO module, as indicated by the Ansible docs. Without `six` (1.1.0), I see `NameError: name 'DoError' is not defined`.

What wasn't initially clear to me is how ansible uses a localhost connection and normally calls the OSX default python interpreter. I also saw this when originally using a brew-installed python, which was always calling OSX interpreter and not the brew one.

To force the virtualenv to be used, I am using this in my `hosts` file:

``` ini
[localhost]
localhost ansible_python_interpreter=python
```

With `six` + the above, the error goes away.

**Comment by reinno** [comment]
I found the same problem on UBUNTU 14.04 ansible 2.0.0.2 and dopy 0.3.7, and 0.3.5 will be ok.

**Comment by davidjpeacock** [comment]
Confirmed reproduced on Mac OS X El Capitan..

Dopy 0.3.7 is broken.
Adding `dopy` 0.3.5 and `six`, and the hack provided by @CraigMason, cured things.

**Comment by frob** [comment]
I have confirmed that this is an issue with the currently released Ansible 2.0.0.2

I have also installed dopy 0.3.5 and it has solved the issue. 

@CraigMason's comment tells us how to select our python environment by setting a variable in the host (this is how ansible know which python has dopy 0.3.5 available to it. His comment shouldn't be related to this issue otherwise.

**Comment by declanmarks** [comment]
Follow this guide on Digital Ocean: https://www.digitalocean.com/community/tutorials/how-to-use-the-digitalocean-api-v2-with-ansible-2-0-on-ubuntu-14-04

This guide states that dopy that the lastest version of dopy is broken and doesn't work with ansible
You need to install 0.3.5 like the person said above.

**Comment by Kralian** [comment]
[on ubuntu 15.10]
Set the following in requirements in a virtualenv and looks like it's running..

ansible==2.0.1.0
dopy==0.3.5
ecdsa==0.13
Jinja2==2.8
MarkupSafe==0.23
paramiko==1.16.0
pycrypto==2.6.1
PyYAML==3.11
requests==1.0.4
six==1.10.0

(before the requests==1.0.4 the error still appeared)

**Comment by geerlingguy** [comment]
Just wanted to post another :+1: here; `sudo pip install 'dopy>=0.3.5,<=0.3.5'` did the trick.

**Comment by robwilkerson** [comment]
This is still kicking my arse. I'm running OS X El Capitan. I've installed python + pip via Homebrew and I've tried installing Ansible via both Homebrew and pip. I'm still getting the error. What am I missing?

> An exception occurred during task execution. To see the full traceback, use -vvv. The error was: NameError: name 'DoError' is not defined

``` shell
$ which python
/usr/local/bin/python
$ which pip
/usr/local/bin/pip
$ which ansible
/usr/local/bin/ansible
$ python --version
Python 2.7.11
$ pip --version
pip 8.1.1 from /usr/local/lib/python2.7/site-packages (python 2.7)
$ ansible --version
ansible 2.0.1.0
  config file = /Users/robwilkerson/Development/uw/ansible-servers/ansible.cfg
  configured module search path = Default w/o overrides
$ pip list
...
dopy (0.3.5)
...
```

**Comment by memoryleak** [comment]
@robwilkerson:
Execute what @geerlingguy posted:

`sudo pip install 'dopy>=0.3.5,<=0.3.5'`

**Comment by robwilkerson** [comment]
@memoryleak That's what I installed (see the truncated `pip list` above). I have something funky happening on my machine. I walked someone else through the same steps I took to arrive at the output you see above and it worked fine for them. I'll have to figure out what's going on, but can't imagine what it would be right now.

**Comment by frob** [comment]
@robwilkerson maybe it is a different bug.

**Comment by robwilkerson** [comment]
@frob Perhaps. It sure feels like the same problem, though, but apparently specific to my environment. I'll follow up here if/when I manage to run it down.

**Comment by ghost** [comment]
Having the same issue as the OP. In my case I have an issue with this shell wrapper for Digitalocean:

https://gist.github.com/richardskumat/df3e789c6f859c886970282fa594a50f

It was setup and used to work well in v2.0.1.0 on Debian Testing, however a dist-upgrade to 2.0.2.0 borked it.

Edit:

I've installed dopy again with pip.

I've pasted the OP's line:

`pip install 'dopy>=0.3.5,<=0.3.5'`

and it started working.

This doesn't make any sense since dopy 0.3.5 was earlier installed on 2.0.1.0 when I was playing with that shell script above and https://www.digitalocean.com/community/tutorials/how-to-use-the-digitalocean-api-v2-with-ansible-2-0-on-ubuntu-14-04.

**Comment by regentgal** [comment]
I no longer receive the error with ansible 2.0.2.0, dopy 0.3.7, and six 1.10.0 on OSX 10.10.5

**Comment by oolongtea** [comment]
Running `pip install dopy` fixed it for me as it was not installed in my virtualenv. No longer receive error on Ubuntu 16.04 LTS with following versions:

```
ansible==2.0.2.0
dopy==0.3.7
six==1.10.0
```

**Comment by frob** [comment]
I think this can be closed

**Comment by schainks** [comment]
Just ran into this via google, confirming @CraigMason's hack works on Mac OS 10.11.4 with 

```
ansible==2.0.2.0
dopy==0.3.7
six==1.10.0
```

**Comment by anshprat** [comment]
Unrelated to this issue but mentioning here since the given workaround causes another issue... dopy 0.3.5 causes private_networking to fail. So use 0.3.4 instead for enabling create time private_networking. Though in 0.3.4, well, only ipv6 fails :/

https://github.com/ansible/ansible-modules-core/issues/1412

**Comment by atulhm** [comment]
I'm stuck.  I still have this error on osx.

ansible (2.0.2.0)
dopy (0.3.5)
httplib2 (0.9)
Jinja2 (2.8)
paramiko (2.0.0)
pip (8.1.2)
six (1.10.0)

An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/Users/djsa;lfjk/.ansible/tmp/ansible-tmp-1465841227.3-278325667029611/digital_ocean", line 185, in <module>
    class TimeoutError(DoError):
NameError: name 'DoError' is not defined

**Comment by atulhm** [comment]
Ok all.  Not sure what is up with my system but I tracked down this obscure issue.  in case anyone else doesn't want to burn half a day....

I have tack on the following to my ansible commands when running with a local connection...

-e 'ansible_python_interpreter=/usr/local/bin/python'

 If anyone has suggestions for what I need to fix on my system, please let me know.  Not sure why ansible isn't finding the right python during local connection tasks.

**Comment by kontrafiktion** [comment]
@atulhm do you have multiple Pythons installed? OS X normally comes with a Python at `/usr/bin/python` and Ansible uses that Python. Try if you can use dopy with /usr/bin/python.
It looks like you are having another python installed at '/usr/local/bin/python' (I assume via HomeBrew?). 
But when Ansible "connects" to localhost it again uses the default interpreter at /usr/bin/python

**Comment by ansibot** [comment]
@alukovenko, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by alukovenko** [comment]
getting rid of dopy in #4219 
wontfix

**Comment by sfenvironment** [comment]
@alukovenko Am to understand by https://github.com/ansible/ansible-modules-core/issues/2509#issuecomment-236551283 that Ansible's core digitalocean module no longer has a requirement of dopy?

Can you please add a bit more information such "as of version xxxx, dopey is no longer required", etc?  Thanks.

**Comment by gregelin** [comment]
@alukovenko Am to understand by #2509 (comment) that Ansible's core digitalocean module no longer has a requirement of dopy?

Can you please add a bit more information such "as of version xxxx, dopey is no longer required", etc? Thanks.

**Comment by alukovenko** [comment]
Maybe @harneksidhu can shed line on current status?

**Comment by harneksidhu** [comment]
@gregelin We are removing the dopy requirement on this ticket #4219. We have a few pull requests already prepared .. #5278 and #5275. We just need to go over it and merge the code into mainline.

**Comment by gregelin** [comment]
@harneksidhu Thank you for the clarifying details, and for removing the dopy requirement. 

Please let us know the version of Ansible which will no longer require dopy.

**Comment by harneksidhu** [comment]
@gregelin To be honest, I am not confident what version of Ansible will no longer require dopy yet. Once the PR's are merged in, we will have a better idea then! The merging does take a while unfortunately.

**Comment by atulhm** [comment]
Hey guys.  I got buried with so much stuff and neglected how I solved my problem.  Since I had 2 or 3 different user environments that invoke ansible, I took the easy way out and dockerized ansible with our specific settings.

I can't wait to make use of the tags with digital ocean.

Thanks for the support.

**Comment by gregelin** [comment]
@harneksidhu Thanks for update, I understand. Will be glad whenever there digitalocean module with no dependency is available.

**Comment by patran** [comment]
Did the removal of the dependency move forward?

**Comment by digitalist** [comment]
It does work for me. The problem was in pyenv and the way in which ansible does import. Maybe it's related only to FreeBSD, but still, I had to link /usr/bin/python directly to pyenv shims python. (since ansible code is full of linuxisms and there is no /usr/bin/python on FreeBSD

If you're running ansible inside pyenv - open digital_ocean.py and try to  import dopy after __future__. If it throws an exception, there's a problem with pyenv.

also it's a minor bug from dopy-side, @zbal, you see, you check for dopy version, but if there's an error importing any dopy version, the message still speaks about 0.3.2
